openvdb_ax/openvdb_ax/compiler/PointExecutable.h

```diff
@@ -0,0 +1,148 @@
+    void setCreateMissing(const bool flag);
+    /// @return  Whether this executable will generate new point attributes.
+    bool getCreateMissing() const;
+
+    /// @brief  Set the threading grain size. Default is 1. A value of 0 has the
+    ///   effect of disabling multi-threading.
+    /// @param grain The grain size
+    void setGrainSize(const size_t grain);
+    /// @return  The current grain size
+    size_t getGrainSize() const;
+
+    ////////////////////////////////////////////////////////
+
+    // @brief deprecated methods
+    OPENVDB_DEPRECATED void
+    execute(points::PointDataGrid& grid,
+        const std::string* const group,
+        const bool create) const
+    {
+        PointExecutable copy(*this);
+        if (group) copy.setGroupExecution(*group);
+        copy.setCreateMissing(create);
+        copy.execute(grid);
+    }
+
+    OPENVDB_DEPRECATED void
+    execute(points::PointDataGrid& grid,
+        const std::string* const group) const
+    {
+        PointExecutable copy(*this);
+        if (group) copy.setGroupExecution(*group);
+        copy.execute(grid);
+    }
+
+    ////////////////////////////////////////////////////////
+
+    // foward declaration of settings for this executable
+    struct Settings;
+
+private:
+    friend class Compiler;
+    friend class ::TestPointExecutable;
+
+    /// @brief Constructor, expected to be invoked by the compiler. Should not
+    ///   be invoked directly.
+    /// @param context Shared pointer to an llvm:LLVMContext associated with the
+    ///   execution engine
+    /// @param engine Shared pointer to an llvm::ExecutionEngine used to build
+    ///   functions. Context should be the associated LLVMContext
+    /// @param attributeRegistry Registry of attributes accessed by AX code
+    /// @param customData Custom data which will be shared by this executable.
+    ///   It can be used to retrieve external data from within the AX code
+    /// @param functions A map of function names to physical memory addresses
+    ///   which were built by llvm using engine
+    PointExecutable(const std::shared_ptr<const llvm::LLVMContext>& context,
+                    const std::shared_ptr<const llvm::ExecutionEngine>& engine,
+                    const AttributeRegistry::ConstPtr& attributeRegistry,
+                    const CustomData::ConstPtr& customData,
+                    const std::unordered_map<std::string, uint64_t>& functions);
+
+private:
+    // The Context and ExecutionEngine must exist _only_ for object lifetime
+    // management. The ExecutionEngine must be destroyed before the Context
+    const std::shared_ptr<const llvm::LLVMContext> mContext;
+    const std::shared_ptr<const llvm::ExecutionEngine> mExecutionEngine;
+    const AttributeRegistry::ConstPtr mAttributeRegistry;
+    const CustomData::ConstPtr mCustomData;
+    const std::unordered_map<std::string, uint64_t> mFunctionAddresses;
+    std::unique_ptr<Settings> mSettings;
```

Comment:
What's the reason for using this pimpl pattern here for settings - is it an attempt to try not to break ABI? More generally, do we need to worry about ABI for these types of objects. If we are going to attempt to do this, why not put all of this data into the implementation as none of it is used in the public interface?